HighAvailabilityModule.Server.InMemory/InMemoryMembershipServer.cs

```diff
@@ -51,24 +64,49 @@ bool ValidInput()
                     return;
                 }
 
-                this.Current = new HeartBeatEntry(uuid, now);
+                this.Current = new HeartBeatEntry(entryDTO.Uuid, entryDTO.Utype, entryDTO.Uname, now);
+
+                this.CurrentTable[entryDTO.Utype] = this.Current;
             }
         }
 
-        public Task<HeartBeatEntry> GetHeartBeatEntryAsync() => this.GetHeartBeatEntryAsync(DateTime.UtcNow);
+        public Task<HeartBeatEntry> GetHeartBeatEntryAsync(string utype) => this.GetHeartBeatEntryAsync(utype, DateTime.UtcNow);
 
-        public async Task<HeartBeatEntry> GetHeartBeatEntryAsync(DateTime now)
+        public async Task<HeartBeatEntry> GetHeartBeatEntryAsync(string utype, DateTime now)
         {
-            if (this.HeartbeatInvalid(now))
+            if (this.HeartbeatInvalid(utype, now))
             {
                 return HeartBeatEntry.Empty;
             }
             else
             {
-                return this.Current;
+                return this.CurrentTable[utype];
+            }                
+        }
+
+        private bool HeartbeatInvalid(string utype, DateTime now)
+        {
+            if (this.CurrentTable.ContainsKey(utype))
+            {
+                return this.CurrentTable[utype] == null || (now - this.CurrentTable[utype].TimeStamp >= this.Timeout);
+            }
+            else
+            {
+                return true;
             }
         }
 
-        private bool HeartbeatInvalid(DateTime now) => this.Current == null || (now - this.Current.TimeStamp >= this.Timeout);
+        private bool LastSeenEntryValid(string utype, HeartBeatEntry LastSeenEntry)
+        {
```

Comment:
`return this.CurrentTable.ContainsKey(utype) && 
    LastSeenEntry != null && 
    this.CurrentTable[utype].Uuid == LastSeenEntry.Uuid &&
    this.CurrentTable[utype].Utype == LastSeenEntry.Utype && 
    this.CurrentTable[utype].TimeStamp == LastSeenEntry.TimeStamp;`